portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -14,19 +14,107 @@
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.appengine.api.datastore.FetchOptions;
+import com.google.appengine.api.datastore.*;
+import com.google.gson.Gson;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
+
 /** Servlet that returns some example content. TODO: modify this file to handle comments data */
-@WebServlet("/data")
+@WebServlet("/add-comment")
 public class DataServlet extends HttpServlet {
 
-  @Override
-  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello world!</h1>");
+    private ArrayList<Message> messages;
+
+    @Override
+    public void init() {
+        this.messages = new ArrayList<Message>();
+    }
+
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {  
+        
+        messages.clear();
+
+        //Original query that works
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        Query query = new Query("Message").addSort("timestamp", SortDirection.DESCENDING);
+        Iterable<Entity> results = datastore.prepare(query).asIterable(FetchOptions.Builder.withLimit(4));
+        
+
+        /* New attempted Solution */
+        // DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        // Query query = new Query("Message").addSort("timestamp", SortDirection.DESCENDING);
+        // PreparedQuery preparedQuery = datastore.prepare(query);
+        // FetchOptions options = FetchOptions.Builder.withLimit(4);
+        // Cursor cursor = preparedQuery.asQueryResultList(options).getCursor();
+        // String encodedCursor = cursor.toWebSafeString();
+        // Pass this encodedCursor back to HTML and save it as global variable in forum.js.
+
+
+        // String startCursor = req.getParameter(request, "comment-cursor-input", "");
+        // if (startCursor == null) {
+            // Cursor cursor = preparedQuery.asQueryResultList(withLimit(4)).getCursor();
+            // String encodedCursor = cursor.toWebSafeString();
+        // }
+        
+        // System.out.println("Encoded cursor is: " + encodedCursor);
+
+        for (Entity messageEntity : results) {
+            long id = messageEntity.getKey().getId();
+            String messageBody = (String) messageEntity.getProperty("body"); 
+            long timestamp = (long) messageEntity.getProperty("timestamp");
+
+            Message newMessage = new Message(id, messageBody, timestamp);
+            messages.add(newMessage);
+        };
```

Comment:
consider using a javastream:
https://www.geeksforgeeks.org/stream-in-java/
https://www.baeldung.com/java-maps-streams
https://stackify.com/streams-guide-java-8/

something like.
```
ArrayList<Messages> messages = results.stream()
    .map(messageEntity -> {
        long id = messageEntity.getKey().getId();
        String messageBody = (String) messageEntity.getProperty("body"); 
        long timestamp = (long) messageEntity.getProperty("timestamp");

        return new Message(id, messageBody, timestamp); 
    }).collect(Collectors.toList());
```